### ISSUE TITLE
Plotting several patches

### ISSUE THREAD

**Issue by pablonoever** [matlab2tikz/matlab2tikz#981]
Hey guys,

I have an issue transfering several patches with matlab2tikz. I have a figure with one axes in Matlab. This axes contains several patches as children that create a shape. It was created with the hold on command by adding a patch after another into the axes. It is in 3D space.
When I run matlab2tikz, the created figure only shows the first of the hundreds of patches. 
Do I miss something?

Thanks for your help and best wishes,

Pablo

**Comment by pablonoever** [comment]
This is the produced file. I reduced the number of patches to three. I tried deleting the 'forget plot'-command and it worked. As it is hard to follow the code in matlab2tikz and I am not too much into PGF plots I do not see the reason, but probably you may see anything.

`% This file was created by matlab2tikz.
%
%The latest updates can be retrieved from
%  http://www.mathworks.com/matlabcentral/fileexchange/22022-matlab2tikz-matlab2tikz
%where you can also make suggestions and rate matlab2tikz.
%
\documentclass[tikz]{standalone}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{pgfplots}
\usepackage{grffile}
\pgfplotsset{compat=newest}
\usetikzlibrary{plotmarks}
\usetikzlibrary{arrows.meta}
\usepgfplotslibrary{patchplots}
\usepackage{amsmath}

% !TeX program = lualatex
\begin{document}
\definecolor{mycolor1}{rgb}{0.55078,0.68750,0.65234}%
%
\begin{tikzpicture}

\begin{axis}[%
width=6.875in,
height=3.209in,
at={(2.675in,0.433in)},
scale only axis,
plot box ratio=2.853 1 1,
xmin=22.9392056926594,
xmax=40.5814287981975,
tick align=outside,
ymin=-4.39948908118591,
ymax=1.78477407193821,
zmin=-2.64714742009965,
zmax=3.53711573302447,
view={15}{15},
axis background/.style={fill=white},
axis x line*=bottom,
axis y line*=left,
axis z line*=left
]

\addplot3[area legend, draw=black, fill=mycolor1, fill opacity=0.8, forget plot]
table[row sep=crcr] {%
x	y	z\\
22.9392056926594	-0.00295897577624008	0.981237639933945\\
22.9392056926594	-0.111949281364459	0.976687414248895\\
22.9392056926594	-0.220865332556338	0.970611278924145\\
22.9392056926594	-0.329696205552291	0.963161895666278\\
22.9392056926594	-0.438435286202925	0.954473639491582\\
22.9392056926594	-0.547079310297836	0.944667181149866\\
22.9392056926594	-0.655627615899425	0.933851875851197\\
22.9392056926594	-0.764082470889232	0.92213593703601\\
22.9392056926594	-0.872448815316907	0.909627308020127\\
22.9392056926594	-0.980732746974124	0.896423950348655\\
22.9392056926594	-1.0889395952831	0.882602859019488\\
22.9392056926594	-1.1970746573681	0.868230745082797\\
22.9392056926594	-1.30514289940667	0.853364410424464\\
22.9392056926594	-1.41314954094245	0.83805684411464\\
22.9392056926594	-1.521100848132	0.822363711475572\\
22.9392056926594	-1.6290034587344	0.806339100466944\\
22.9392056926594	-1.73686327875676	0.790028908094058\\
22.9392056926594	-1.84468495363486	0.77346839543297\\
22.9392056926594	-1.95247158791975	0.756681311948337\\
22.9392056926594	-2.06022499295008	0.739682225588633\\
22.9392056926594	-2.16794989666725	0.722503418772228\\
22.9392056926594	-2.27565213059145	0.705183033082882\\
22.9392056926594	-2.3833362152897	0.687750155519294\\
22.9392056926594	-2.49100524431274	0.670224524016967\\
22.9392056926594	-2.59866296507049	0.652629553472747\\
24.7036454250553	-2.64252688639394	0.583560410354401\\
24.7036454250553	-2.53134324664863	0.600901497684237\\
24.7036454250553	-2.42014842832277	0.618170752224449\\
24.7036454250553	-2.30893872157232	0.635343859718084\\
24.7036454250553	-2.19771085582467	0.652398945088012\\
24.7036454250553	-2.08646016571046	0.669304483217377\\
24.7036454250553	-1.97518045946209	0.686017922673509\\
24.7036454250553	-1.86386787027527	0.702510925492018\\
24.7036454250553	-1.75252039260837	0.718766697299059\\
24.7036454250553	-1.64113454766998	0.734757418061608\\
24.7036454250553	-1.52970515445886	0.75044173827291\\
24.7036454250553	-1.41822585206644	0.765767202202887\\
24.7036454250553	-1.30668999131375	0.780675352800547\\
24.7036454250553	-1.19509166744311	0.795108358491704\\
24.7036454250553	-1.08342598324866	0.809010446748056\\
24.7036454250553	-0.971687690794027	0.822316131830247\\
24.7036454250553	-0.859871647574747	0.83495149376442\\
24.7036454250553	-0.747971972557024	0.846822932646151\\
24.7036454250553	-0.635982943092991	0.857818921667201\\
24.7036454250553	-0.523901281418287	0.867825932589098\\
24.7036454250553	-0.411726883854108	0.876732460758083\\
24.7036454250553	-0.299462196459004	0.884416210886031\\
24.7036454250553	-0.187112996117701	0.890743215649832\\
24.7036454250553	-0.0746890897281672	0.895560195049247\\
24.7036454250553	0.0377942525721809	0.898689984851562\\
}--cycle;

\addplot3[area legend, draw=black, fill=mycolor1, fill opacity=0.8, forget plot]
table[row sep=crcr] {%
x	y	z\\
24.7036454250553	0.0377942525721809	0.898689984851562\\
24.7036454250553	-0.0746890897281676	0.895560195049247\\
24.7036454250553	-0.187112996117701	0.890743215649832\\
24.7036454250553	-0.299462196459005	0.884416210886031\\
24.7036454250553	-0.411726883854108	0.876732460758083\\
24.7036454250553	-0.523901281418287	0.867825932589098\\
24.7036454250553	-0.635982943092991	0.857818921667201\\
24.7036454250553	-0.747971972557024	0.846822932646151\\
24.7036454250553	-0.859871647574747	0.83495149376442\\
24.7036454250553	-0.971687690794027	0.822316131830247\\
24.7036454250553	-1.08342598324866	0.809010446748056\\
24.7036454250553	-1.19509166744311	0.795108358491704\\
24.7036454250553	-1.30668999131375	0.780675352800547\\
24.7036454250553	-1.41822585206644	0.765767202202887\\
24.7036454250553	-1.52970515445887	0.75044173827291\\
24.7036454250553	-1.64113454766998	0.734757418061608\\
24.7036454250553	-1.75252039260837	0.718766697299059\\
24.7036454250553	-1.86386787027527	0.702510925492018\\
24.7036454250553	-1.97518045946209	0.686017922673509\\
24.7036454250553	-2.08646016571046	0.669304483217377\\
24.7036454250553	-2.19771085582467	0.652398945088012\\
24.7036454250553	-2.30893872157232	0.635343859718084\\
24.7036454250553	-2.42014842832277	0.618170752224449\\
24.7036454250553	-2.53134324664863	0.600901497684237\\
24.7036454250553	-2.64252688639394	0.583560410354401\\
26.4680524811155	-2.66612243619263	0.514255351003982\\
26.4680524811155	-2.5531481565399	0.531472896841514\\
26.4680524811155	-2.44016261979288	0.548616408617767\\
26.4680524811155	-2.32716233096741	0.56566240174125\\
26.4680524811155	-2.21414296581981	0.582581432916335\\
26.4680524811155	-2.10109858505328	0.59933247248517\\
26.4680524811155	-1.9880220432039	0.615864974783525\\
26.4680524811155	-1.87490953724527	0.632149577343774\\
26.4680524811155	-1.76175852926165	0.648164436471425\\
26.4680524811155	-1.64856491098393	0.663875233213586\\
26.4680524811155	-1.53532273816959	0.679232054219617\\
26.4680524811155	-1.42202589436113	0.694180086775308\\
26.4680524811155	-1.30866910817891	0.708666427608185\\
26.4680524811155	-1.19524745653657	0.722635669123399\\
26.4680524811155	-1.08175571970636	0.736023373317688\\
26.4680524811155	-0.968189034651597	0.748759414778837\\
26.4680524811155	-0.854543305884342	0.760769327262899\\
26.4680524811155	-0.740813248099735	0.771951936300312\\
26.4680524811155	-0.626994443274637	0.782191106753079\\
26.4680524811155	-0.513085193133335	0.791368587978069\\
26.4680524811155	-0.399086686778315	0.799359957959962\\
26.4680524811155	-0.285003448151489	0.806031663899573\\
26.4680524811155	-0.17084438811927	0.811244598843082\\
26.4680524811155	-0.0566228161048832	0.814830490570021\\
26.4680524811155	0.0576408093095827	0.816622104596463\\
}--cycle;

\addplot3[area legend, draw=black, fill=mycolor1, fill opacity=0.8, forget plot]
table[row sep=crcr] {%
x	y	z\\
31.7610258387061	0.0492860941967368	0.565953220637291\\
31.7610258387061	-0.0635016699609052	0.566292191305862\\
31.7610258387061	-0.176280325429949	0.564803096417286\\
31.7610258387061	-0.289025109648951	0.561657629189495\\
31.7610258387061	-0.401717931055818	0.557005958581281\\
31.7610258387061	-0.514345467217155	0.55097515332794\\
31.7610258387061	-0.626897592997685	0.543669643423836\\
31.7610258387061	-0.739369490785228	0.535217005324573\\
31.7610258387061	-0.851761077617694	0.525755227956771\\
31.7610258387061	-0.964075127459637	0.515412894696959\\
31.7610258387061	-1.07631534208039	0.504297440659856\\
31.7610258387061	-1.18848318409854	0.492473786570206\\
31.7610258387061	-1.30057850377947	0.479981083341075\\
31.7610258387061	-1.41260550102695	0.466889595970279\\
31.7610258387061	-1.52456940760717	0.453268905886184\\
31.7610258387061	-1.63647190181466	0.439152521100276\\
31.7610258387061	-1.74831436007467	0.424568007429737\\
31.7610258387061	-1.86010206804788	0.409569485927911\\
31.7610258387061	-1.97184116690955	0.394212951697795\\
31.7610258387061	-2.08353469023065	0.378528321809251\\
31.7610258387061	-2.19518721262991	0.362554370640156\\
31.7610258387061	-2.30680770890286	0.346358069944304\\
31.7610258387061	-2.41840589282337	0.330008693945951\\
31.7610258387061	-2.52998975155333	0.313561812746596\\
31.7610258387061	-2.6415633720805	0.297045617985625\\
33.5252477824943	-2.61202257559077	0.229094758286506\\
33.5252477824943	-2.50177419930447	0.245709510873557\\
33.5252477824943	-2.39152792249717	0.262338191978144\\
33.5252477824943	-2.28127771676674	0.278940800132664\\
33.5252477824943	-2.1710187687259	0.295485242124914\\
33.5252477824943	-2.0607440472426	0.311924188472702\\
33.5252477824943	-1.95044460054002	0.328196374677817\\
33.5252477824943	-1.84011487925142	0.344261976159301\\
33.5252477824943	-1.7297501937573	0.360085555981022\\
33.5252477824943	-1.6193437978413	0.375615355000429\\
33.5252477824943	-1.50888752105964	0.390786186816496\\
33.5252477824943	-1.39837196537446	0.405518792530777\\
33.5252477824943	-1.28778856449305	0.419733000145734\\
33.5252477824943	-1.17713113132297	0.433358672825178\\
33.5252477824943	-1.06639465984241	0.446326070743607\\
33.5252477824943	-0.955574403793408	0.458556407139269\\
33.5252477824943	-0.84466556212918	0.469954971059426\\
33.5252477824943	-0.733664880437813	0.480420786517175\\
33.5252477824943	-0.622571198876183	0.489847451163197\\
33.5252477824943	-0.511384944906953	0.49810952612924\\
33.5252477824943	-0.400109647809674	0.505068542549376\\
33.5252477824943	-0.288753542485872	0.510582726188347\\
33.5252477824943	-0.177329805611987	0.514494593804181\\
33.5252477824943	-0.0658583376930217	0.516645137772729\\
33.5252477824943	0.0456334155941795	0.516878774728128\\
}--cycle;
\end{axis}
\end{tikzpicture}%
\end{document}`

**Comment by okomarov** [comment]
Ca you provide a Matlab snippet that produces the figure please?

**Comment by pablonoever** [comment]
Hi okomarov,

the problem resolved itself, after I've run the file today again. But I got a different problem:

In the generated plot in latex the patches are not in 3D space, as they schould be. They are shown on top depending on the time when they are plotted. Last is on top, although it is three-dimensionally behind the others for example. Compare matlab and latex plot.

If you have a resolution for that it would be great otherwise I have to use a png file...

For the code below please use the data in the attached file.

`load('Zones.mat');
%width and height of plot
w=30;h=10;

f=figure('Units','centimeters',...
        'Position',[0 0 w h],...
        'PaperPositionMode','auto');
a=axes;

for i=1:length(X)
   patch(a,Z{i},X{i},Y{i},C{i},'FaceAlpha',1,...
         'EdgeColor',[.1 .1 .1],'LineWidth',0.1);
end

hold(a, 'on');
axis(a,'equal');

view(40,20);
matlab2tikz('filename','Zone.tex','floatFormat','%.3f','figurehandle',f,'standalone',true)
close(f)`

[Zones.zip](https://github.com/matlab2tikz/matlab2tikz/files/855147/Zones.zip)

**Comment by okomarov** [comment]
From what I see in Matlab, a surface plot should be more approapriate. Also, the plot is very heavy and it does not compile. If you could provide a simplified example we can dig further.

**Comment by pablonoever** [comment]
Hi okomarov,

thanks for your reply. I've attached a reduced geometry that you can still open with the same skript. The X-Y grid is not regular and so not that easily to plot with surface, different than with patches. Plus depending on the skript behind some of the patches may be visibel or not. So somehow I have to stay with the patches.

[Zones_short.zip](https://github.com/matlab2tikz/matlab2tikz/files/855449/Zones_short.zip)

**Comment by pablonoever** [comment]
Hi okomarov,

have you got time to check my issue, or anyone else? Would be great to find some hints where I failed.

Best wishes,

Pablo
